chaos/net-rpc-client.go

```diff
@@ -26,135 +25,154 @@ import (
-		if err.Error() == rpc.ErrShutdown.Error() {
-			// Reset rpcClient.rpc to nil to trigger a reconnect in future
-			// and close the underlying connection.
-			rpcClient.clearRPCClient()
-
-			// Close the underlying connection.
-			rpcLocalStack.Close()
-
-			// Set rpc error as rpc.ErrShutdown type.
-			err = rpc.ErrShutdown
-		}
+	err := rpcClient.rpc.Call(serviceMethod, args, reply)
+	if IsRPCError(err) {
+		rpcClient.rpc.Close()
+		rpcClient.rpc = nil
 	}
 	return err
 }
 
 func (rpcClient *ReconnectRPCClient) RLock(args dsync.LockArgs) (status bool, err error) {
 	err = rpcClient.Call("Dsync.RLock", &args, &status)
+	if err == rpc.ErrShutdown {
+		if ierr := rpcClient.Init(); ierr != nil {
+			return false, ierr
+		}
+		err = rpcClient.Call("Dsync.RLock", &args, &status)
+	}
 	return status, err
 }
 
 func (rpcClient *ReconnectRPCClient) Lock(args dsync.LockArgs) (status bool, err error) {
 	err = rpcClient.Call("Dsync.Lock", &args, &status)
+	if err == rpc.ErrShutdown {
+		if ierr := rpcClient.Init(); ierr != nil {
+			return false, ierr
+		}
+		err = rpcClient.Call("Dsync.Lock", &args, &status)
+	}
 	return status, err
 }
 
 func (rpcClient *ReconnectRPCClient) RUnlock(args dsync.LockArgs) (status bool, err error) {
 	err = rpcClient.Call("Dsync.RUnlock", &args, &status)
+	if err == rpc.ErrShutdown {
+		if ierr := rpcClient.Init(); ierr != nil {
+			return false, ierr
+		}
+		err = rpcClient.Call("Dsync.RUnlock", &args, &status)
+	}
 	return status, err
 }
 
 func (rpcClient *ReconnectRPCClient) Unlock(args dsync.LockArgs) (status bool, err error) {
 	err = rpcClient.Call("Dsync.Unlock", &args, &status)
+	if err == rpc.ErrShutdown {
+		if ierr := rpcClient.Init(); ierr != nil {
+			return false, ierr
+		}
+		err = rpcClient.Call("Dsync.Unlock", &args, &status)
+	}
 	return status, err
 }
 
 func (rpcClient *ReconnectRPCClient) ForceUnlock(args dsync.LockArgs) (status bool, err error) {
 	err = rpcClient.Call("Dsync.ForceUnlock", &args, &status)
+	if err == rpc.ErrShutdown {
+		if ierr := rpcClient.Init(); ierr != nil {
+			return false, ierr
+		}
+		err = rpcClient.Call("Dsync.ForceUnlock", &args, &status)
+	}
 	return status, err
 }
 
-// Close closes the underlying socket file descriptor.
-func (rpcClient *ReconnectRPCClient) Close() error {
-	// See comment above for making a copy on local stack
-	rpcLocalStack := rpcClient.getRPCClient()
-
-	// If rpc client has not connected yet there is nothing to close.
-	if rpcLocalStack == nil {
-		return nil
-	}
+func (rpcClient *ReconnectRPCClient) ServerAddr() string {
+	return rpcClient.addr
+}
 
-	// Reset rpcClient.rpc to allow for subsequent calls to use a new
-	// (socket) connection.
-	rpcClient.clearRPCClient()
-	return rpcLocalStack.Close()
+func (rpcClient *ReconnectRPCClient) ServiceEndpoint() string {
+	return rpcClient.endpoint
 }
 
-func (rpcClient *ReconnectRPCClient) ServerAddr() string {
-	return rpcClient.node
+func (rpcClient *ReconnectRPCClient) Init() error {
```

Comment:
There is no need for NetLocker method specific retry logic.
1. In dsync, the retry on failure is not the responsibility of the implementation of NetLocker.
2. In this implementation of NetLocker, there is only one single retry behaviour. So, we don't need it.
So, it's an avoidable repetition.